### PR TITLE
Initial React framework setup, sidebar, hash routing, registration tab.

### PR DESCRIPTION
**Pull request by kyh8**

### CODE REVIEW

File: src/main/webapp/css/main.css [hunk 1]
```diff
@@ -0,0 +1,272 @@
+body {
```
Comment (jiaweizhang) [inline src/main/webapp/css/main.css:1]: Do we plan on ever separating this into multiple css files?

File: src/main/webapp/js/App.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    return (
+      <div>
+        <div id="content-container">
+          <Router history={hashHistory}>
+            <Route path="/~kevinhe/dpc" component={Content}>
```
Comment (jiaweizhang) [inline src/main/webapp/js/App.js]: Why is there a /~kevinhe in the path lol?

File: src/main/webapp/js/Content.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  render() {
+    return (
+      <div>
+        <div id="sidebar-container">
+          <Sidebar />
+        </div>
+        <div className="main-content" style={{
+          width: (this.state.dimensions.width - 60) + 'px',
```
Comment (jiaweizhang) [inline src/main/webapp/js/Content.js:67]: any way we can move this into the CSS?

File: src/main/webapp/index.html [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
-    <meta charset="UTF-8">
-    <title>Duke Programming Contest</title>
-</head>
-<body>
-Duke Programming Contest
-</body>
-</html>
\ No newline at end of file
+<html>
+  <head>
```
Comment (jiaweizhang) [inline src/main/webapp/index.html:3]: let's look into using google analytics to track accesses and use. I'll make an issue separately.

File: src/main/webapp/js/Content.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        dimensions: this.calculateDimensions()
+      });
+    }
+
+    this.state = {
+      dimensions: this.calculateDimensions()
+    }
+  }
+
+  calculateDimensions() {
```
Comment (jiaweizhang) [inline src/main/webapp/js/Content.js:44]: Can we move these hardcoded values to CSS somehow?

File: src/main/webapp/js/Home.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  render() {
+    const width = this.props.dimensions.width - 160;
+    const height = this.props.dimensions.height - 100;
+
+    return (
+      <div className="home-tab">
+        <div
+          className="home-intro card"
+          style={{
+            left: '50px',
```
Comment (jiaweizhang) [inline src/main/webapp/js/Home.js]: belongs in css

File: src/main/webapp/js/Problems.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    number: 2,
+    points: 5,
+    difficulty: 'Medium',
+    title: 'Knapsack Problem'
+  },
+  {
+    number: 3,
+    points: 5,
+    difficulty: 'Easy',
+    title: 'My Google Inteview Question'
```
Comment (jiaweizhang) [inline src/main/webapp/js/Problems.js:19]: change this before Google sues us.

File: src/main/webapp/js/Registration.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  submitLoginInfo(e) {
+    e.preventDefault();
+    console.log('submitted');
+    const emailField = document.getElementById('email-login-field');
+    const passwordField = document.getElementById('password-login-field');
+    const email = emailField.value;
+    const password = passwordField.value;
+    console.log(email, ',', password);
+    if (email.length == 0 || password.length == 0) {
```
Comment (jiaweizhang) [inline src/main/webapp/js/Registration.js]: will create issue to do more front-end validation

File: src/main/webapp/js/Registration.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
+const React = require('react');
+const TITLE_HEIGHT = 100;
```
Comment (jiaweizhang) [inline src/main/webapp/js/Registration.js]: Is there a way to move this to CSS and use css-classes to retrieve these values?

File: src/main/webapp/js/Sidebar.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    sidebarTabs.forEach((tab, index) => {
+      elements.push(
+        <Link key={tab.name + "-tab"} to={"/" + (tab.name.toLowerCase() !== "home" ? tab.name.toLowerCase() : "")}>
+          <div
+            className={
+              this.state.selected === index
+              ? "sidebar-tab selected"
+              : "sidebar-tab"
+            }
+            onClick={this.selectTab.bind(this, index)}
```
Comment (jiaweizhang) [inline src/main/webapp/js/Sidebar.js]: Isn't this deprecated?

File: src/main/webapp/js/Util.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+export class Util {
+  static calculateDimensions(xOffset, yOffset) {
```
Comment (jiaweizhang) [inline src/main/webapp/js/Util.js:2]: Why do we have "calculateDimensions" all over the place?

File: src/main/webapp/package.json [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  "name": "dpc",
+  "version": "1.0.0",
+  "description": "Interface for Duke's Programming Contest.",
+  "main": "main.js",
+  "scripts": {
+    "watch": "watchify -t [ babelify ] main.js -o bundle.js -v",
+    "build": "browserify -t [ babelify ] main.js  | uglifyjs -mc warnings=false > bundle.js"
+  },
+  "author": "Kevin He",
+  "license": "ISC",
```
Comment (jiaweizhang) [inline src/main/webapp/package.json]: change to GPL 3.0

File: src/main/webapp/package.json [hunk 13]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "dpc",
+  "version": "1.0.0",
+  "description": "Interface for Duke's Programming Contest.",
+  "main": "main.js",
+  "scripts": {
+    "watch": "watchify -t [ babelify ] main.js -o bundle.js -v",
+    "build": "browserify -t [ babelify ] main.js  | uglifyjs -mc warnings=false > bundle.js"
```
Comment (jiaweizhang) [inline src/main/webapp/package.json:8]: does uglifyjs perform tree-balancing and minimization?

File: src/main/webapp/index.html [hunk 14]
```diff
@@ -1,10 +1,17 @@
...
+    <meta charset="UTF-8" />
+    <title> Duke Programming Contest </title>
+    <link href="css/main.css" rel="stylesheet">
+    <link href="font-awesome/css/font-awesome.min.css" rel="stylesheet">
+    <link rel="icon" href="assets/code-brackets.png">
```
Comment (jiaweizhang) [inline src/main/webapp/index.html:8]: I assume none of these images/icons are copyrighted and we can use?

**Review by jiaweizhang — CHANGES_REQUESTED** [review]

**Comment by jiaweizhang** [comment]
Also, let's remove bundle.js and add instructions on how others can run the front-end and how they can access it via localhost

File: src/main/webapp/css/main.css (hunk 1 above)
Comment (kyh8) [reply src/main/webapp/css/main.css:1]: Yes, eventually if we need to. Right now just putting it in one file

File: src/main/webapp/js/App.js (hunk 2 above)
Comment (kyh8) [reply src/main/webapp/js/App.js]: oops

File: src/main/webapp/js/Content.js (hunk 5 above)
Comment (kyh8) [reply src/main/webapp/js/Content.js:44]: It has to be done on resize, which is a js thing

File: src/main/webapp/js/Content.js (hunk 3 above)
Comment (kyh8) [reply src/main/webapp/js/Content.js:67]: Same as above

File: src/main/webapp/index.html (hunk 14 above)
Comment (kyh8) [reply src/main/webapp/index.html:8]: Yes

File: src/main/webapp/js/Problems.js (hunk 7 above)
Comment (kyh8) [reply src/main/webapp/js/Problems.js:19]: This is dummy data

File: src/main/webapp/js/Registration.js (hunk 8 above)
Comment (kyh8) [reply src/main/webapp/js/Registration.js]: This is not the final thing lel

File: src/main/webapp/js/Sidebar.js (hunk 10 above)
Comment (kyh8) [reply src/main/webapp/js/Sidebar.js]: onClick?

File: src/main/webapp/package.json (hunk 13 above)
Comment (kyh8) [reply src/main/webapp/package.json:8]: Dunno dude

File: src/main/webapp/js/Content.js (hunk 5 above)
Comment (jiaweizhang) [reply src/main/webapp/js/Content.js:44]: But can't you do like "update to the value contained by this class"

**Comment by kyh8** [comment]
Please merge

**Review by jiaweizhang — APPROVED** [review]

**Pull request merged**